### PR TITLE
[Framework] Always initialize port default mappings in case where no other mapping is configured

### PR DESCRIPTION
**Pull request by omby8888**
# Description

What - Always initialize port default mappings in case where no other mapping is configured
Why - Integration that initializes with empty mapping wont do enything in resync and will cause confusion with our clients, furthermore integration that runs 'ONCE' will take no effect with initializeportresources=false
How - reorder the initialization proccess

## Type of change

Please leave one option from the following and delete the rest:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] New Integration (non-breaking change which adds a new integration)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] Non-breaking change (fix of existing functionality that will not change current behavior)
- [ ] Documentation (added/updated documentation)

## Screenshots

Include screenshots from your environment showing how the resources of the integration will look.

## API Documentation

Provide links to the API documentation used for this integration.

### CODE REVIEW

File: port_ocean/core/defaults/initialize.py [hunk 1]
```diff
@@ -50,10 +53,47 @@ def deconstruct_blueprints_to_creation_steps(
...
+                port_app_config=default_mapping,
+            )
+        elif not integration["config"]:
+            await port_client.patch_integration(
+                integration_config.integration.type,
+                integration_config.event_listener.to_request(),
+                port_app_config=default_mapping,
+            )
+    except httpx.HTTPStatusError as err:
+        raise Exception(f"Failed to apply default mapping: {err.response.text}.")
```
Comment (yairsimantov20) [inline port_ocean/core/defaults/initialize.py:76]: Please just log and raise the same exception

File: port_ocean/run.py [hunk 2]
```diff
@@ -49,9 +49,8 @@ def run(
     if initialize_port_resources is not None:
         app.config.initialize_port_resources = initialize_port_resources
 
-    if app.config.initialize_port_resources:
```
Comment (yairsimantov20) [inline port_ocean/run.py:52]: Why did u remove feature flags?

File: port_ocean/core/defaults/initialize.py [hunk 3]
```diff
@@ -205,12 +246,9 @@ async def _initialize_defaults(
...
 ) -> None:
-    try:
-        asyncio.new_event_loop().run_until_complete(
-            _initialize_defaults(config_class, integration_config)
-        )
-    except Exception as e:
-        logger.debug(f"Failed to initialize defaults, skipping... Error: {e}")
+    asyncio.new_event_loop().run_until_complete(
+        _initialize_integration(config_class, integration_config)
+    )
```
Comment (yairsimantov20) [inline port_ocean/core/defaults/initialize.py:254]: Why did u femove exception handling?

File: port_ocean/core/defaults/initialize.py [hunk 4]
```diff
@@ -163,18 +197,25 @@ async def _create_resources(
...
-        logger.warning("No defaults found. Skipping...")
+        logger.warning("No defaults found. Skipping initialization...")
         return None
 
+    await _initialize_required_integration_settings(
+        port_client, defaults.port_app_config, integration_config
+    )
+
+    if not integration_config.initialize_port_resources:
+        return
```
Comment (yairsimantov20) [inline port_ocean/core/defaults/initialize.py:214]: What if i expect the integration to fail when starting up if does not exist 
Should we handle this?

File: port_ocean/core/defaults/initialize.py [hunk 5]
```diff
@@ -163,18 +197,25 @@ async def _create_resources(
         )
 
 
-async def _initialize_defaults(
+async def _initialize_integration(
```
Comment (yairsimantov20) [inline port_ocean/core/defaults/initialize.py:200]: Why did you rename? You are changing much more under the scope of this function

File: port_ocean/core/defaults/initialize.py [hunk 6]
```diff
@@ -50,10 +53,47 @@ def deconstruct_blueprints_to_creation_steps(
...
+            changelog_destination,
+            default_mapping,
+        )
+
+
 async def _create_resources(
     port_client: PortClient,
     defaults: Defaults,
-    integration_config: IntegrationConfiguration,
 ) -> None:
```
Comment (yairsimantov20) [inline port_ocean/core/defaults/initialize.py:97]: What about trying to create some of the missing blueprints?

**Review by yairsimantov20 — CHANGES_REQUESTED** [review]
Few comments
also you are missing changelog and version bump
please bump it by minor and not by patch like usual i want to start managing the versions normally

File: port_ocean/core/defaults/initialize.py (hunk 6 above)
Comment (omby8888) [reply port_ocean/core/defaults/initialize.py:97]: is that in scope of this issue?

File: port_ocean/run.py (hunk 2 above)
Comment (omby8888) [reply port_ocean/run.py:52]: because then it wont enter the method and skipp the whole initializtion
i moved this if into the function

File: port_ocean/core/defaults/initialize.py (hunk 3 above)
Comment (omby8888) [reply port_ocean/core/defaults/initialize.py:254]: don't we want it to fail in case the mapping couldn't be applied? i think it will confuse the user if we'll continue normally and resyncs wouldn't work

File: port_ocean/core/defaults/initialize.py (hunk 4 above)
Comment (omby8888) [reply port_ocean/core/defaults/initialize.py:214]: not sure i got what u meant

File: port_ocean/core/defaults/initialize.py [hunk 7]
```diff
@@ -73,7 +73,8 @@ async def _initialize_required_integration_settings(
                 port_app_config=default_mapping,
             )
     except httpx.HTTPStatusError as err:
-        raise Exception(f"Failed to apply default mapping: {err.response.text}.")
+        logger.error("Failed to apply default mapping")
```
Comment (yairsimantov20) [inline port_ocean/core/defaults/initialize.py:76]:
```suggestion
        logger.error(f"Failed to apply default mapping: {err.response.text}.")
```

File: port_ocean/core/defaults/initialize.py [hunk 8]
```diff
@@ -205,12 +248,9 @@ async def _initialize_defaults(
...
+def initialize_integration(
     config_class: Type[PortAppConfig], integration_config: IntegrationConfiguration
 ) -> None:
-    try:
-        asyncio.new_event_loop().run_until_complete(
-            _initialize_defaults(config_class, integration_config)
-        )
-    except Exception as e:
-        logger.debug(f"Failed to initialize defaults, skipping... Error: {e}")
+    asyncio.new_event_loop().run_until_complete(
```
Comment (yairsimantov20) [inline port_ocean/core/defaults/initialize.py:254]: we have a file in the cli folder that use the create defaults make sure you are not breaking itz

**Review by yairsimantov20 — CHANGES_REQUESTED** [review]
more comments

File: port_ocean/core/defaults/initialize.py (hunk 6 above)
Comment (yairsimantov20) [reply port_ocean/core/defaults/initialize.py:98]: Please talk to @MPTG94

File: port_ocean/core/defaults/initialize.py (hunk 7 above)
Comment (omby8888) [reply port_ocean/core/defaults/initialize.py:76]: that was the previous state, don't think it's needed to both log the error text and raise the whole exception

File: port_ocean/core/defaults/initialize.py (hunk 8 above)
Comment (omby8888) [reply port_ocean/core/defaults/initialize.py:255]: yeah it works

File: port_ocean/core/defaults/initialize.py (hunk 6 above)
Comment (omby8888) [reply port_ocean/core/defaults/initialize.py:99]: he said that in any case if one or more of the blueprints that the integration needs to create in its initialization already exist, he expects the integration to not create any blueprints (just keep the state as it is), so no need for that for now at least

File: CHANGELOG.md [hunk 9]
```diff
@@ -7,6 +7,14 @@ this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 
 <!-- towncrier release notes start -->
 
+## 0.6.0 (2024-06-09)
+
+
+### Improvements
+
+- Changed initialization to always apply default mapping if no other mapping is configured
```
Comment (Tankilevitch) [inline CHANGELOG.md:15]: add the case that you added for run once

File: port_ocean/core/defaults/initialize.py [hunk 10]
```diff
@@ -50,18 +53,52 @@ def deconstruct_blueprints_to_creation_steps(
...
+            should_log=False, should_raise=False
+        )
+        if not integration:
+            integration = await port_client.create_integration(
+                integration_config.integration.type,
+                integration_config.event_listener.to_request(),
+                port_app_config=default_mapping,
+            )
+        elif not integration["config"]:
+            integration = await port_client.patch_integration(
```
Comment (Tankilevitch) [inline port_ocean/core/defaults/initialize.py:73]: add log about each case, indicating what you got and what you are about to apply

File: port_ocean/core/defaults/initialize.py [hunk 11]
```diff
@@ -50,18 +53,52 @@ def deconstruct_blueprints_to_creation_steps(
...
+    if (
+        integration["changelogDestination"] != changelog_destination
+        or integration["installationAppType"] != integration_config.integration.type
+        or integration.get("version") != port_client.integration_version
+    ):
+        await port_client.patch_integration(
+            integration_config.integration.type,
+            changelog_destination,
+            default_mapping,
+        )
```
Comment (Tankilevitch) [inline port_ocean/core/defaults/initialize.py:95]: why is this? I don't understand why we would like to override the integration config if there is a diff

**Review by Tankilevitch — CHANGES_REQUESTED** [review]

File: port_ocean/core/defaults/initialize.py (hunk 11 above)
Comment (omby8888) [reply port_ocean/core/defaults/initialize.py:95]: i copied it from `port_ocean/clients/port/mixins/integrations.py`, not sure why is that exactly but i dont wanna change the current behavior regarding this, maybe @yairsimantov20 knows more

File: port_ocean/core/defaults/initialize.py (hunk 7 above)
Comment (yairsimantov20) [reply port_ocean/core/defaults/initialize.py:84]: The whole exception wont log the response text

File: port_ocean/core/defaults/initialize.py (hunk 11 above)
Comment (yairsimantov20) [reply port_ocean/core/defaults/initialize.py:100]: @omby8888 you can omit the default mapping from there
Just update the version, type and changelog destination

**Review by Tankilevitch — APPROVED** [review]
LGTM

**Review by yairsimantov20 — APPROVED** [review]
🌊 dont forget to change the changelog date
